### PR TITLE
Correct kernel metadata for LFRic adjoint code (closes #1772)

### LINKED ISSUE #1772: modify Adjoint metadata
Metadata names and access patterns may need to be updated when transforming from tangent-linear metadata to adjoint metadata. When dealing with stencil codes further modications may be required. At the moment this is not easily done as the PSyIR captures the metadata as an UnknownFortranType. Work is ongoing to provide a PSyIR interface to the metadata. Once this is done the metadata can be updated.

### PR DESCRIPTION
**Pull request by rupertford**
Generated LFRic adjoint code needs the original tangent-linear kernel metadata modifying. For example, the access properties may change e.g. GH_READ might become GH_INC. This PR does not modify the metadata for mixed precision kernels - that will be another PR. There are also more complicated cases involving changes to the argument list but these are not solved here either.  With these changes, 10 LFRIC adjoint kernels can be generated automatically (with no tweaks) from their tangent-linear kernels.

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for first review. Another for @arporter I'm afraid as it is psyad metadata functionality.

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (arporter) [inline src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:367]: This is good but needs a bit of explaining. Is this 'subroutine arg: corresponding metadata arg'?

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (arporter) [inline src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:364]: Could you include the kernel name too if possible?

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (arporter) [inline src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:373]: I was a bit worried about this but forgot to comment earlier. How does this sit with the rule that only builtins are allowed to have a scalar argument that is written to?

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (arporter) [inline src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:469]: "intent in"?

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (arporter) [inline src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:499]: Kernerl name pls if poss.

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (arporter) [inline src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:601]: s/is also added//

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (arporter) [inline src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:83]: Have you removed this as we no longer support multiple kernel routines in a file?

**Review by arporter — CHANGES_REQUESTED** [review]
Nice. It's really good to be able to do this now.
Just minor things.
Ref Guide still builds OK.
I've set the integration tests running.

File: examples/psyad/lfric/tangent_linear/Makefile
Comment (rupertford) [reply examples/psyad/lfric/tangent_linear/Makefile:57]: Agreed

File: examples/psyad/lfric/tangent_linear/Makefile
Comment (rupertford) [reply examples/psyad/lfric/tangent_linear/Makefile:50]: No, because these are the files that now produce correct adjoint code from tangent-linear code so can go straight into the adjoint directory, however, there are other files in the adjoint directory that have had manual changes.

File: examples/psyad/lfric/tangent_linear/Makefile
Comment (rupertford) [reply examples/psyad/lfric/tangent_linear/Makefile:59]: Added some words

Comment (rupertford) [reply examples/psyad/lfric/tangent_linear/Makefile:59]: Actually I ended up removing them as the information did not seem to be in the right place IMO.

File: examples/psyad/lfric/tangent_linear_tweaked/Makefile
Comment (rupertford) [reply examples/psyad/lfric/tangent_linear_tweaked/Makefile:67]: added.

File: examples/psyad/lfric/tangent_linear_tweaked/matrix_vector_kernel_mod_tweaked.F90
Comment (rupertford) [reply examples/psyad/lfric/tangent_linear_tweaked/matrix_vector_kernel_mod_tweaked.F90:33]: Updated all tweaked kernels with comments about the tweaks.

File: examples/psyad/lfric/tangent_linear_tweaked/dg_inc_matrix_vector_kernel_mod_tweaked.F90
Comment (rupertford) [reply examples/psyad/lfric/tangent_linear_tweaked/dg_inc_matrix_vector_kernel_mod_tweaked.F90:34]: Updated all tweaked kernels with comments about the tweaks.

File: src/psyclone/domain/lfric/utils.py
Comment (rupertford) [reply src/psyclone/domain/lfric/utils.py:83]: Updated.

File: src/psyclone/domain/lfric/utils.py
Comment (rupertford) [reply src/psyclone/domain/lfric/utils.py:116]: Oops.

File: src/psyclone/domain/lfric/utils.py
Comment (rupertford) [reply src/psyclone/domain/lfric/utils.py:127]: As you mention later, I don't think so. `_mod_mod` would be matched, as would `_mmoodd`.

Comment (rupertford) [reply src/psyclone/domain/lfric/utils.py:127]: But I have tidied a little

File: src/psyclone/psyad/domain/lfric/lfric_adjoint.py
Comment (rupertford) [reply src/psyclone/psyad/domain/lfric/lfric_adjoint.py:259]: "By jove he's got it."

File: src/psyclone/psyad/domain/lfric/lfric_adjoint.py
Comment (rupertford) [reply src/psyclone/psyad/domain/lfric/lfric_adjoint.py:258]: Modified.

File: src/psyclone/tests/domain/lfric/test_utils.py
Comment (rupertford) [reply src/psyclone/tests/domain/lfric/test_utils.py:54]: Oops.

File: src/psyclone/tests/domain/lfric/test_utils.py
Comment (rupertford) [reply src/psyclone/tests/domain/lfric/test_utils.py:117]: oops

File: src/psyclone/tests/domain/lfric/test_utils.py
Comment (rupertford) [reply src/psyclone/tests/domain/lfric/test_utils.py:132]: cut-n-paste error. updated.

File: src/psyclone/tests/domain/lfric/test_utils.py
Comment (rupertford) [reply src/psyclone/tests/domain/lfric/test_utils.py:166]: Updated error message :-)

File: src/psyclone/tests/domain/lfric/test_utils.py
Comment (rupertford) [reply src/psyclone/tests/domain/lfric/test_utils.py:175]: Extended test

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (rupertford) [reply src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:367]: Attempted to improve explanation.

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (rupertford) [reply src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:364]: No the way it is written as the method `_update_access_metadata` is not given this information. I could re-raise outside of `_update_access_metadata` and add the kernel name, although PSyAD only works on one kernel at a time anyway.

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (rupertford) [reply src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:373]: Good question. I guess I could wibble about using on builtins, or asking if it is the job of PSyAD (mine is not to reason why) but I'm not sure they are valid answers. If this were to be flagged (and exception raise in PSyAD) then I thimk it would make sense to do so when working out the intents of the active variables in the adjoint code, rather than here so perhaps I should create an issue and mark it here as something that needs changing - or pehaps I should raise an exception here too and change the test? I'll defer to you on what you think is best.

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (rupertford) [reply src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:469]: No, that is correct, it is just that the docstring is badly written. I've hopefully improved it enough.

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (rupertford) [reply src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:601]: Oops.

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (rupertford) [reply src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:83]: Yes, when using the new metadata processing approach it does not currently support multi-precision kernels so I dropped the multi-kernel test. I am working towards multi-precision support, it is just that the metadata values currently disappear in PSyIR so I need to work on that first before tackling the fact that metadata multi-precision kernels are treated as structures not unknown fortran type (as they don't have a contains/procedure), so need parsing separately.

**Comment by rupertford** [comment]
Ready for next review from @arporter

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (arporter) [reply src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py]: Thanks.

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (arporter) [reply src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:377]: Since this needs to go on I suggest creating a new issue and adding TODOs both here and in the code that works out the intents of the active variables.

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (arporter) [inline src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:365]: Please rm this line.

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (arporter) [reply src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:83]: OK.

**Review by arporter — CHANGES_REQUESTED** [review]
Just a few minor things to fix and a new Issue and TODOs needed.
Ref. guide builds without new warnings.
You need to update the UG now that multiple subroutines are not supported (https://psyclone-adjoint.readthedocs.io/en/latest/implementation.html#multiple-subroutines). In fact, we should probably have a "metadata" section under "Implementation".
Integration tests were previously green, I'll check these again next time.

**Comment by arporter** [comment]
Just realised I was in the middle of tidying up some code so taking this back temporarily.

File: src/psyclone/tests/domain/lfric/test_utils.py
Comment (arporter) [reply src/psyclone/tests/domain/lfric/test_utils.py]: I've taken the liberty of fixing this as it's just a test docstring.

File: src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py
Comment (arporter) [reply src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint.py:377]: I've created #2333 and added TODOs here and in `lfric_adjoint::_update_access_metadata`.

File: examples/psyad/lfric/tangent_linear_tweaked/Makefile
Comment (arporter) [reply examples/psyad/lfric/tangent_linear_tweaked/Makefile:70]: Done.

**Comment by arporter** [comment]
Since this is the last PR needed prior to release and my requested changes were only docstrings and documentation, I've taken the liberty of doing them. This therefore needs a quick look from either @hiker or @TeranIvy as I can't legitimately merge it to master now.

**Review by TeranIvy — COMMENTED** [review]
Tests with compilation pass and examples run. Developer and user guide build and display nicely.

I also built the PSyAD user guide as it was changed by this PR. Everything is fine, however I noticed that the `doc/psyad/user_guide/_build/` directory was displayed as untracked changes, so I took the opportunity to edit the main `.gitignore` file to not track this.

Updated the user guide and changelog for merge to master. Awaiting the CI to complete before proceeding with the merge.

**Review by TeranIvy — APPROVED** [review]
CI passed, proceeding to merge.

**Review by TeranIvy — APPROVED** [review]

**Pull request merged**
